pipert/core/message.py

```diff
@@ -0,0 +1,107 @@
+from abc import ABC, abstractmethod
+from io import BytesIO
+
+import numpy as np
+import time
+import pickle
+import cv2
+
+from PIL import Image
+
+
+class Payload(ABC):
+
+    def __init__(self, data):
+        self.data = data
+        pass
+
+    @abstractmethod
+    def decode(self):
+        pass
+
+    @abstractmethod
+    def encode(self):
+        pass
+
+
+class FramePayload(Payload):
+
+    def __init__(self, data):
+        super().__init__(data)
+
+    def decode(self):
+        self.data = np.array(Image.open(BytesIO(self.data)))
+
+    def encode(self):
+        _, buf = cv2.imencode(".jpg", self.data)
+        self.data = buf.tobytes()
+
+
+class PredictionPayload(Payload):
+    def __init__(self, data):
+        super().__init__(data)
+
+    def decode(self):
+        pass
+
+    def encode(self):
+        pass
+
+
+class Message:
+    next_msg_id = 0
+
+    def __init__(self, data, stream_address):
+        if isinstance(data, np.ndarray):
+            self.payload = FramePayload(data)
+        else:
+            self.payload = PredictionPayload(data)
+        self.stream_address = stream_address
+        self.history = {}
+        self.id = Message.next_msg_id
+        Message.next_msg_id += 1
+
+    def update_payload(self, data):
+        self.payload.data = data
+
+    def get_payload(self):
+        return self.payload.data
+
+    # component name should represent a unique instance of the component
+    def record_entry(self, component_name):
+        self.history[component_name] = [time.time(), None]
+
+    def record_exit(self, component_name):
+        if component_name in self.history:
+            self.history[component_name][1] = time.time()
+        else:
+            self.history[component_name] = [None, time.time()]
```

Comment:
Maybe store times into dictionary and not list? In case someone want to record a custom timestamp for a given component / routine or something.

Reply:
Can you explain what you mean? Having trouble understanding this one @ItamarWilf

Reply:
Right now, for each record you add a timestamp to a list. Right now that list will have a size of 2, so you can access the timestamps by index. Let's assume sometime in the future someone will want to record the entry to some pre processing, so he will add another record to the list. In order to support that and still have a readable history, maybe we should store the records in a dict with informative keys.

Reply:
Maybe I'm missing something, but it already works as you described - each message holds a dictionary called 'history', which maps a string to a pair of time stamps (one represents the entry, the other represents the exit). So if we want to record the entry to some pre processing we can just do the following: `msg.record_entry(component_name + "_preprocessing")`. And when it exits from the preprocessing we can do `msg.record_exit(component_name + "_preprocessing")`
Meaning each message can already contain a history of when it entered and exited a particular section of the pipeline. 
@ItamarWilf

Reply:
Your'e correct, but let's say we have a detection stage and a post processing stage. Recording the exit from the recording stage and then recording the entry to the post processing stage is redundant. It's not that important but just something that gives us a little more control and clarity for monitoring later on.